### ISSUE TITLE
Handler executed multiple times when using include

### ISSUE THREAD

**Issue by bitonio** [ansible/ansible#21608]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

handlers

##### ANSIBLE VERSION

```
2.2.1.0
```

##### OS / ENVIRONMENT

N/A

##### SUMMARY
Handlers with include statement are called multipled times (as many times as notify).

##### STEPS TO REPRODUCE
```
- name: Testing notify handler with include
  hosts: blue
  become: True
  tasks:
    - command: id
      notify: restart test
    - command: id
      notify: restart test
    - command: id
      notify: restart test
    - command: id
      notify: restart test
  handlers:
    - name: restart test
      include: tasks/restart_test.yml
```

And the restart_test.yml:

```
- name: Restart test (we use NTP)
  service: name=ntp state=restarted
```

Output
```
ansible-playbook test-restart.yml

PLAY [Testing notify handler with include] *************************************

TASK [setup] *******************************************************************
ok: [host01]

TASK [command] *****************************************************************
changed: [host01]

TASK [command] *****************************************************************
changed: [host01]

TASK [command] *****************************************************************
changed: [host01]

TASK [command] *****************************************************************
changed: [host01]

RUNNING HANDLER [Restart test (we use NTP)] ************************************
changed: [host01]
changed: [host01]
changed: [host01]
changed: [host01]

PLAY RECAP *********************************************************************
host01                     : ok=9    changed=8    unreachable=0    failed=0
```

##### EXPECTED RESULTS

Handler should be executed only once.

##### ACTUAL RESULTS

Handler being call as many times as notify

**Comment by alikins** [comment]
Can reproduce this with devel/ branch.

test.yml:

```
- name: Testing notify handler with include
  hosts: localhost
  #become: True
  gather_facts: false
  tasks:
    - command: id
      notify: "restart test debug"
    - command: id
      notify: "restart test debug"
    - command: id
      notify: "restart test debug"
    - command: id
      notify: "restart test debug"
  handlers:
    - name: restart test debug
      include: tasks/restart_test_debug.yml
```

tasks/restart_test_debug.yml
```
- name: Restart test (simulated just a debug)
  debug:
    msg: "restart handler called"
```

output  (the command used in tasks changed to echo, but otherwise the same):

```
[fedora-25:ansible (devel % u=)]$ ansible-playbook -vvv -i hosts repro/handlers_multiple_calls_21608/test.yml 
Using /home/adrian/ansible/ansible.cfg as config file
statically included: /home/adrian/src/ansible/repro/handlers_multiple_calls_21608/tasks/restart_test_debug.yml

PLAYBOOK: test.yml *******************************************************************************************************************************************************************************************************************************************
1 plays in repro/handlers_multiple_calls_21608/test.yml

PLAY [Testing notify handler with include] *******************************************************************************************************************************************************************************************************************
META: ran handlers

TASK [command] ***********************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible/repro/handlers_multiple_calls_21608/test.yml:5
Using module file /home/adrian/src/ansible/lib/ansible/modules/commands/command.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604615.57-164149914039738 `" && echo ansible-tmp-1487604615.57-164149914039738="` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604615.57-164149914039738 `" ) && sleep 0'                                                                                                                                                                                                                                           
<localhost> PUT /tmp/tmpc7VxLB TO /home/adrian/.ansible/tmp/ansible-tmp-1487604615.57-164149914039738/command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1487604615.57-164149914039738/ /home/adrian/.ansible/tmp/ansible-tmp-1487604615.57-164149914039738/command.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1487604615.57-164149914039738/command.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1487604615.57-164149914039738/" > /dev/null 2>&1 && sleep 0'
NOTIFIED HANDLER Restart test (simulated just a debug)
changed: [localhost] => {
    "changed": true, 
    "cmd": [
        "echo"
    ], 
    "delta": "0:00:00.002026", 
    "end": "2017-02-20 10:30:15.690069", 
    "invocation": {
        "module_args": {
            "_raw_params": "echo", 
            "_uses_shell": false, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }
    }, 
    "rc": 0, 
    "start": "2017-02-20 10:30:15.688043", 
    "stderr": "", 
    "stderr_lines": [], 
    "stdout": "", 
    "stdout_lines": []
}

TASK [command] ***********************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible/repro/handlers_multiple_calls_21608/test.yml:7
Using module file /home/adrian/src/ansible/lib/ansible/modules/commands/command.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604615.72-80637427890430 `" && echo ansible-tmp-1487604615.72-80637427890430="` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604615.72-80637427890430 `" ) && sleep 0'                                                                                                                                                                                                                                              
<localhost> PUT /tmp/tmpIJ5PI3 TO /home/adrian/.ansible/tmp/ansible-tmp-1487604615.72-80637427890430/command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1487604615.72-80637427890430/ /home/adrian/.ansible/tmp/ansible-tmp-1487604615.72-80637427890430/command.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1487604615.72-80637427890430/command.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1487604615.72-80637427890430/" > /dev/null 2>&1 && sleep 0'
NOTIFIED HANDLER Restart test (simulated just a debug)
changed: [localhost] => {
    "changed": true, 
    "cmd": [
        "echo"
    ], 
    "delta": "0:00:00.002112", 
    "end": "2017-02-20 10:30:15.840965", 
    "invocation": {
        "module_args": {
            "_raw_params": "echo", 
            "_uses_shell": false, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }
    }, 
    "rc": 0, 
    "start": "2017-02-20 10:30:15.838853", 
    "stderr": "", 
    "stderr_lines": [], 
    "stdout": "", 
    "stdout_lines": []
}

TASK [command] ***********************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible/repro/handlers_multiple_calls_21608/test.yml:9
Using module file /home/adrian/src/ansible/lib/ansible/modules/commands/command.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604615.87-271400875433406 `" && echo ansible-tmp-1487604615.87-271400875433406="` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604615.87-271400875433406 `" ) && sleep 0'                                                                                                                                                                                                                                           
<localhost> PUT /tmp/tmpWvTFgk TO /home/adrian/.ansible/tmp/ansible-tmp-1487604615.87-271400875433406/command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1487604615.87-271400875433406/ /home/adrian/.ansible/tmp/ansible-tmp-1487604615.87-271400875433406/command.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1487604615.87-271400875433406/command.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1487604615.87-271400875433406/" > /dev/null 2>&1 && sleep 0'
NOTIFIED HANDLER Restart test (simulated just a debug)
changed: [localhost] => {
    "changed": true, 
    "cmd": [
        "echo"
    ], 
    "delta": "0:00:00.002090", 
    "end": "2017-02-20 10:30:15.988779", 
    "invocation": {
        "module_args": {
            "_raw_params": "echo", 
            "_uses_shell": false, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }
    }, 
    "rc": 0, 
    "start": "2017-02-20 10:30:15.986689", 
    "stderr": "", 
    "stderr_lines": [], 
    "stdout": "", 
    "stdout_lines": []
}

TASK [command] ***********************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible/repro/handlers_multiple_calls_21608/test.yml:11
Using module file /home/adrian/src/ansible/lib/ansible/modules/commands/command.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604616.02-177566930724191 `" && echo ansible-tmp-1487604616.02-177566930724191="` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604616.02-177566930724191 `" ) && sleep 0'
<localhost> PUT /tmp/tmphasaJV TO /home/adrian/.ansible/tmp/ansible-tmp-1487604616.02-177566930724191/command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1487604616.02-177566930724191/ /home/adrian/.ansible/tmp/ansible-tmp-1487604616.02-177566930724191/command.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1487604616.02-177566930724191/command.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1487604616.02-177566930724191/" > /dev/null 2>&1 && sleep 0'
NOTIFIED HANDLER Restart test (simulated just a debug)
changed: [localhost] => {
    "changed": true, 
    "cmd": [
        "echo"
    ], 
    "delta": "0:00:00.001871", 
    "end": "2017-02-20 10:30:16.139542", 
    "invocation": {
        "module_args": {
            "_raw_params": "echo", 
            "_uses_shell": false, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }
    }, 
    "rc": 0, 
    "start": "2017-02-20 10:30:16.137671", 
    "stderr": "", 
    "stderr_lines": [], 
    "stdout": "", 
    "stdout_lines": []
}

RUNNING HANDLER [Restart test (simulated just a debug)] ******************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "restart handler called"
}
ok: [localhost] => {
    "changed": false, 
    "msg": "restart handler called"
}
ok: [localhost] => {
    "changed": false, 
    "msg": "restart handler called"
}
ok: [localhost] => {
    "changed": false, 
    "msg": "restart handler called"
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=8    changed=4    unreachable=0    failed=0   

```

**Comment by alikins** [comment]
Note that the behavior changes depending if the include statement in the handler is 'static' or not:

test_static_true.yml
``` yaml
- name: Testing notify handler with include
  hosts: localhost
  #become: True
  gather_facts: false
  tasks:
    - command: echo
      notify: "restart test debug"
    - command: echo
      notify: "restart test debug"
  handlers:
    - name: restart test debug
      include: tasks/restart_test_debug.yml
      static: true
```

output:

```
[fedora-25:ansible (devel % u=)]$ ansible-playbook -vvv -i hosts repro/handlers_multiple_calls_21608/test_static_true.yml 
Using /home/adrian/ansible/ansible.cfg as config file
statically included: /home/adrian/src/ansible/repro/handlers_multiple_calls_21608/tasks/restart_test_debug.yml

PLAYBOOK: test_static_true.yml *******************************************************************************************************************************************************************************************************************************
1 plays in repro/handlers_multiple_calls_21608/test_static_true.yml

PLAY [Testing notify handler with include] *******************************************************************************************************************************************************************************************************************
META: ran handlers

TASK [command] ***********************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible/repro/handlers_multiple_calls_21608/test_static_true.yml:6
Using module file /home/adrian/src/ansible/lib/ansible/modules/commands/command.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604923.82-134541335685050 `" && echo ansible-tmp-1487604923.82-134541335685050="` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604923.82-134541335685050 `" ) && sleep 0'                                                                                                                                                                                                                                           
<localhost> PUT /tmp/tmp0AeUzR TO /home/adrian/.ansible/tmp/ansible-tmp-1487604923.82-134541335685050/command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1487604923.82-134541335685050/ /home/adrian/.ansible/tmp/ansible-tmp-1487604923.82-134541335685050/command.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1487604923.82-134541335685050/command.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1487604923.82-134541335685050/" > /dev/null 2>&1 && sleep 0'
NOTIFIED HANDLER Restart test (simulated just a debug)
changed: [localhost] => {
    "changed": true, 
    "cmd": [
        "echo"
    ], 
    "delta": "0:00:00.002036", 
    "end": "2017-02-20 10:35:23.940972", 
    "invocation": {
        "module_args": {
            "_raw_params": "echo", 
            "_uses_shell": false, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }
    }, 
    "rc": 0, 
    "start": "2017-02-20 10:35:23.938936", 
    "stderr": "", 
    "stderr_lines": [], 
    "stdout": "", 
    "stdout_lines": []
}

TASK [command] ***********************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible/repro/handlers_multiple_calls_21608/test_static_true.yml:8
Using module file /home/adrian/src/ansible/lib/ansible/modules/commands/command.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604923.97-39167348025384 `" && echo ansible-tmp-1487604923.97-39167348025384="` echo /home/adrian/.ansible/tmp/ansible-tmp-1487604923.97-39167348025384 `" ) && sleep 0'
<localhost> PUT /tmp/tmpUOgCKP TO /home/adrian/.ansible/tmp/ansible-tmp-1487604923.97-39167348025384/command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1487604923.97-39167348025384/ /home/adrian/.ansible/tmp/ansible-tmp-1487604923.97-39167348025384/command.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1487604923.97-39167348025384/command.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1487604923.97-39167348025384/" > /dev/null 2>&1 && sleep 0'
NOTIFIED HANDLER Restart test (simulated just a debug)
changed: [localhost] => {
    "changed": true, 
    "cmd": [
        "echo"
    ], 
    "delta": "0:00:00.001809", 
    "end": "2017-02-20 10:35:24.087446", 
    "invocation": {
        "module_args": {
            "_raw_params": "echo", 
            "_uses_shell": false, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }
    }, 
    "rc": 0, 
    "start": "2017-02-20 10:35:24.085637", 
    "stderr": "", 
    "stderr_lines": [], 
    "stdout": "", 
    "stdout_lines": []
}

RUNNING HANDLER [Restart test (simulated just a debug)] ******************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "restart handler called"
}
ok: [localhost] => {
    "changed": false, 
    "msg": "restart handler called"
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=0   
```


With static: false, it handler only runs once:

test_static_false.yml:
``` yaml
- name: Testing notify handler with include
  hosts: localhost
  #become: True
  gather_facts: false
  tasks:
    - command: echo
      notify: "restart test debug"
    - command: echo
      notify: "restart test debug"
  handlers:
    - name: restart test debug
      include: tasks/restart_test_debug.yml
      static: false
```

output:

```
[fedora-25:ansible (devel % u=)]$ ansible-playbook -vvv -i hosts repro/handlers_multiple_calls_21608/test_static_false.yml 
Using /home/adrian/ansible/ansible.cfg as config file

PLAYBOOK: test_static_false.yml ******************************************************************************************************************************************************************************************************************************
1 plays in repro/handlers_multiple_calls_21608/test_static_false.yml

PLAY [Testing notify handler with include] *******************************************************************************************************************************************************************************************************************
META: ran handlers

TASK [command] ***********************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible/repro/handlers_multiple_calls_21608/test_static_false.yml:6
Using module file /home/adrian/src/ansible/lib/ansible/modules/commands/command.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1487605022.51-251200380958380 `" && echo ansible-tmp-1487605022.51-251200380958380="` echo /home/adrian/.ansible/tmp/ansible-tmp-1487605022.51-251200380958380 `" ) && sleep 0'                                                                                                                                                                                                                                           
<localhost> PUT /tmp/tmp2w0Los TO /home/adrian/.ansible/tmp/ansible-tmp-1487605022.51-251200380958380/command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1487605022.51-251200380958380/ /home/adrian/.ansible/tmp/ansible-tmp-1487605022.51-251200380958380/command.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1487605022.51-251200380958380/command.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1487605022.51-251200380958380/" > /dev/null 2>&1 && sleep 0'
NOTIFIED HANDLER restart test debug
changed: [localhost] => {
    "changed": true, 
    "cmd": [
        "echo"
    ], 
    "delta": "0:00:00.002171", 
    "end": "2017-02-20 10:37:02.633362", 
    "invocation": {
        "module_args": {
            "_raw_params": "echo", 
            "_uses_shell": false, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }
    }, 
    "rc": 0, 
    "start": "2017-02-20 10:37:02.631191", 
    "stderr": "", 
    "stderr_lines": [], 
    "stdout": "", 
    "stdout_lines": []
}

TASK [command] ***********************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible/repro/handlers_multiple_calls_21608/test_static_false.yml:8
Using module file /home/adrian/src/ansible/lib/ansible/modules/commands/command.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1487605022.66-233462715708659 `" && echo ansible-tmp-1487605022.66-233462715708659="` echo /home/adrian/.ansible/tmp/ansible-tmp-1487605022.66-233462715708659 `" ) && sleep 0'
<localhost> PUT /tmp/tmpRgzox1 TO /home/adrian/.ansible/tmp/ansible-tmp-1487605022.66-233462715708659/command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1487605022.66-233462715708659/ /home/adrian/.ansible/tmp/ansible-tmp-1487605022.66-233462715708659/command.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1487605022.66-233462715708659/command.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1487605022.66-233462715708659/" > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true, 
    "cmd": [
        "echo"
    ], 
    "delta": "0:00:00.001755", 
    "end": "2017-02-20 10:37:02.783313", 
    "invocation": {
        "module_args": {
            "_raw_params": "echo", 
            "_uses_shell": false, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }
    }, 
    "rc": 0, 
    "start": "2017-02-20 10:37:02.781558", 
    "stderr": "", 
    "stderr_lines": [], 
    "stdout": "", 
    "stdout_lines": []
}

RUNNING HANDLER [restart test debug] *************************************************************************************************************************************************************************************************************************
included: /home/adrian/src/ansible/repro/handlers_multiple_calls_21608/tasks/restart_test_debug.yml for localhost

RUNNING HANDLER [debug] **************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "restart handler called"
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=0   
```

**Comment by alikins** [comment]
@jimi-c @bcoca Looks like this is a bug. Feel free to reassign as appropriate.

**Comment by jimi-c** [comment]
This is a duplicate of https://github.com/ansible/ansible/issues/18178, so we'll track it there.
